File: solutions/049-group-anagrams/solution.go
package main

import (
	"fmt"
	"sort"
)

func groupAnagrams(strs []string) [][]string {
	groups := make(map[[26]int][]string)

	for _, s := range strs {
		var count [26]int
		for i := 0; i < len(s); i++ {
			count[s[i]-'a']++
		}
		groups[count] = append(groups[count], s)
	}

	ans := make([][]string, 0, len(groups))
	for _, g := range groups {
		ans = append(ans, g)
	}
	return ans
}

func normalize(groups [][]string) [][]string {
	out := make([][]string, len(groups))
	for i, g := range groups {
		cp := append([]string(nil), g...)
		sort.Strings(cp)
		out[i] = cp
	}
	sort.Slice(out, func(i, j int) bool {
		return out[i][0] < out[j][0]
	})
	return out
}

func main() {
	tests := []struct {
		strs     []string
		expected [][]string
	}{
		{
			[]string{"eat", "tea", "tan", "ate", "nat", "bat"},
			[][]string{{"bat"}, {"nat", "tan"}, {"ate", "eat", "tea"}},
		},
		{[]string{""}, [][]string{{""}}},
		{[]string{"a"}, [][]string{{"a"}}},
	}

	for _, tc := range tests {
		result := groupAnagrams(tc.strs)
		ok := fmt.Sprint(normalize(result)) == fmt.Sprint(normalize(tc.expected))
		status := "PASS"
		if !ok {
			status = "FAIL"
		}
		fmt.Printf("%s | strs=%v => %v\n", status, tc.strs, result)
	}
}

File: solutions/049-group-anagrams/solution.py
from collections import defaultdict
from typing import List


class Solution:
    def group_anagrams(self, strs: List[str]) -> List[List[str]]:
        groups: dict[tuple[int, ...], list[str]] = defaultdict(list)

        for s in strs:
            count = [0] * 26
            for ch in s:
                count[ord(ch) - ord("a")] += 1
            groups[tuple(count)].append(s)

        return list(groups.values())


def normalize(groups: List[List[str]]) -> List[List[str]]:
    return sorted(sorted(g) for g in groups)


if __name__ == "__main__":
    tests = [
        (
            ["eat", "tea", "tan", "ate", "nat", "bat"],
            [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]],
        ),
        ([""], [[""]]),
        (["a"], [["a"]]),
    ]
    sol = Solution()
    for strs, expected in tests:
        result = sol.group_anagrams(strs)
        ok = normalize(result) == normalize(expected)
        status = "PASS" if ok else "FAIL"
        print(f"{status} | strs={strs!r} => {result}")
